File: src/main/progression.ts
export const POUR_INTERVAL_MS = 15 * 60 * 1000;
export const REFERENCE_X = 0.13;

export interface ProgressState {
  countedMs: number;
  extraWineCount: number;
  waterCount: number;
  processedPourInterval: number;
  encouraged: boolean;
  celebrated: boolean;
}

export function markerX(state: ProgressState): number {
  const sipEquivalent = Math.max(0, state.countedMs / POUR_INTERVAL_MS + state.extraWineCount - state.waterCount);
  return REFERENCE_X * sipEquivalent / 8;
}

export function curveY(x: number): number {
  return Math.log1p(x / REFERENCE_X) / Math.log(2);
}

export function applyCountedTime<T extends ProgressState>(state: T, countedMs: number): T {
  const next = { ...state, countedMs: Math.max(state.countedMs, countedMs) };
  const interval = Math.floor(next.countedMs / POUR_INTERVAL_MS);
  next.processedPourInterval = Math.max(state.processedPourInterval, interval);
  next.encouraged = state.encouraged || next.countedMs >= 30 * 60 * 1000;
  next.celebrated = state.celebrated || markerX(next) >= REFERENCE_X;
  return next;
}

export function giveWine<T extends ProgressState>(state: T): T {
  const next = { ...state, extraWineCount: state.extraWineCount + 1 };
  next.celebrated = state.celebrated || markerX(next) >= REFERENCE_X;
  return next;
}

export function drinkWater<T extends ProgressState>(state: T): T {
  return { ...state, waterCount: state.waterCount + 1 };
}

File: Sources/TiddlyCore/Progression.swift
import Foundation

/// Pure progression model. Counted work time and wine move the marker forward; water moves it back one sip.
public enum Progression {
  public static let pourIntervalMs: Double = 15 * 60 * 1000
  public static let referenceX: Double = 0.13
  public static let encouragementMs: Double = 30 * 60 * 1000
  public static let sleepyMs: Double = 5 * 60 * 60 * 1000

  public static func markerX(_ state: PetState) -> Double {
    let sips = max(0, state.countedMs / pourIntervalMs + Double(state.extraWineCount) - Double(state.waterCount))
    return referenceX * sips / 8
  }

  public static func curveY(_ x: Double) -> Double { log1p(x / referenceX) / log(2) }

  public static func nextPourMs(_ state: PetState) -> Double {
    pourIntervalMs - state.countedMs.truncatingRemainder(dividingBy: pourIntervalMs)
  }

  public static func applyCountedTime(_ state: PetState, _ countedMs: Double) -> PetState {
    var next = state
    next.countedMs = max(state.countedMs, countedMs)
    next.processedPourInterval = max(state.processedPourInterval, Int(floor(next.countedMs / pourIntervalMs)))
    next.encouraged = state.encouraged || next.countedMs >= encouragementMs
    next.celebrated = state.celebrated || markerX(next) >= referenceX
    return next
  }

  public static func giveWine(_ state: PetState) -> PetState {
    var next = state
    next.extraWineCount += 1
    next.celebrated = state.celebrated || markerX(next) >= referenceX
    return next
  }

  public static func drinkWater(_ state: PetState) -> PetState {
    var next = state
    next.waterCount += 1
    return next
  }
}
